### PR TITLE
Fix --freeze-all-layers for new model builder path + unit tests

### PR DESCRIPTION
**Pull request by AAnoosheh**
- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?
Fix the freeze_all_layers arg now that there's a model builder construction path which ignored it before, and allow it to skip checkpoint saving when enabled (+ unit tests)

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact @NVIDIA/mcore-oncall.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [x] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [x] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [x] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

### CODE REVIEW

**Comment by AAnoosheh** [comment]
/ok to test [74f3ff1](https://github.com/NVIDIA/Megatron-LM/pull/5926/commits/74f3ff1bd1e15cced8c6c7eb155cd85685c510ba)

**Comment by AAnoosheh** [comment]
/ok to test [62ec28f](https://github.com/NVIDIA/Megatron-LM/pull/5926/commits/62ec28f27321cb76ea6b249ce1fdd501efb098d4)

File: megatron/training/training.py [hunk 1]
```diff
@@ -2358,20 +2382,21 @@ def train_step(forward_step_func, data_iterator, model, optimizer, opt_param_sch
...
+        with _forward_backward_grad_context(args):
+            losses_reduced = forward_backward_func(
+                forward_step_func=forward_step_func,
+                data_iterator=data_iterator,
+                model=model,
+                num_microbatches=get_num_microbatches(),
+                seq_length=args.seq_length,
+                micro_batch_size=args.micro_batch_size,
+                decoder_seq_length=args.decoder_seq_length,
+                forward_only=False,
```
Comment (jenchen13) [inline megatron/training/training.py]: forward_only should be toggled based on `--freeze-all-layers`, there is no point to allowing backward if you are freezing layers

Comment (AAnoosheh) [reply megatron/training/training.py]: Luckily `schedules.backward_step()` skips backward if the output tensor doesn't require grad. But I agree, would be still good to do as you suggest, in case that behavior changes.

**Comment by AAnoosheh** [comment]
/ok to test [611ac4b](https://github.com/NVIDIA/Megatron-LM/pull/5926/commits/611ac4bd668528c4cd9308f90036c680274d4143)

File: megatron/training/checkpointing.py [hunk 2]
```diff
@@ -531,6 +549,12 @@ def save_checkpoint(iteration, model, optimizer, opt_param_scheduler, num_floati
     # Only rank zero of the data parallel writes to the disk.
     model = unwrap_model(model)
 
+    # --freeze-all-layers: the model weights are frozen and never change, so there is nothing new
```
Comment (jenchen13) [inline megatron/training/checkpointing.py]: nit: this comment can be shorter

File: megatron/training/checkpointing.py [hunk 3]
```diff
@@ -2061,6 +2106,33 @@ def load_checkpoint(ddp_model, optimizer, opt_param_scheduler, load_arg='load',
     else:
         print_rank_0('could not find arguments in the checkpoint ...')
 
+    # --override-ckpt-iteration: rewind the data loader to this iteration, operating directly on
+    # `args` (not the loaded state_dict) so it also works on checkpoints without a saved `args`
+    # blob -- e.g. release or HuggingFace-converted checkpoints, where the state_dict path would
+    # silently no-op. The GBS-match guard applies only when this checkpoint's own args are being
```
Comment (jenchen13) [inline megatron/training/checkpointing.py]: this comment is too long

File: megatron/training/training.py [hunk 4]
```diff
@@ -2058,6 +2084,7 @@ def _build_model_wrapper(wrap_with_ddp: bool):
         )
 
     model = _build_model_wrapper(wrap_with_ddp)
+
```
Comment (jenchen13) [inline megatron/training/training.py]: remove this?

**Review by jenchen13 — APPROVED** [review]
LGTM

**Review by jaredcasper — APPROVED** [review]

**Review by cuichenx — APPROVED** [review]

**Comment by AAnoosheh** [comment]
/ok to test [4e53ddd](https://github.com/NVIDIA/Megatron-LM/pull/5926/commits/4e53ddd8b626d647fa78f61028935e4e74ae981e)
